### DISCUSSION TITLE
Support OCI Registry For Code Upload

### DISCUSSION THREAD

**Discussion by manno** [epinio/epinio#1103] [Ideas]
The current implementation uploads the user's code to an S3-compatible object store, i.e. minio.

OCI registries support arbitrary data. 
This could allow Epinio to drop minio and use the same registry for source code and images. 

The main problem seems to be that [dockerhub doesn’t support the media type](https://github.com/docker/roadmap/issues/135). Epinio could use a, OCI spec-breaking, workaround, like [cosign](https://github.com/sigstore/cosign/issues/913) does.

There are [go libraries](https://golangrepo.com/repo/oras-project-oras-go-go-utilities) to deal with uploading and downloading.

If Epinio wants to have more plugability, it could support both S3 and OCI registries. When running locally, e.g. Rancher Desktop the OCI registry might be sufficient. Large production deployments probably want to use S3.

* Do we need to support all registries?
* How ephemeral are the code blobs?

**Comment by jimmykarily** [comment]
Putting some links here that I found useful:

- Registries that support this: https://oras.land/implementors/
- Go library for oras: https://oras.land/client_libraries/go/

In my opinion, we should switch to OCI artifacts and not support S3 unless someone has a good reason and asks for it. Given most of the public cloud providers support artifacts on their registries and that our registry can support it too, I don't see why someone would want to use S3. Maybe only because of pricing (S3 based solutions may be cheaper than registries when it comes to storage and data transfers).

Switching allows us to get rid of one additional dependency in Epinio, will all the setup and complexity it comes with.

@agracey what do you think?

**Reply by agracey** [reply]
I like the idea. I have two concerns:

- Would we ever want to support the ability to transfer diffs? IIRC, the s3 protocol allows for patching files without transferring the whole blob.
- Is there any performance or size limitations that we would hit? I assume no since I know SUSE in the past has produced and stored 3-4GB OCI images 🤣

If neither of those are an issue, I'd say go for it. The simplicity is really nice.

**Comment by manno** [comment]
As we don't have any features implemented regarding the uploaded code blobs, now would be the time.
Well not entirely true, I think we can re-stage the currently running app, but we do not have history, versioning, etc.

The main decision here for me is, do we assume:
* users want to bring their own code blob store, and we want to support as many as possible, to match all users (plugability)?
* or this is just an implementation detail and one protocol is enough. Users still have the choice which type of OCI registry they want to run and where to run it.

Just supporting OCI registries sounds more sane to me. When we do  versioning and history, we can use the same protocol for uploaded sources, images, etc.

**Reply by agracey** [reply]
If we monotonically increase the tag, we can do rollback/restage with OCI.

**Reply by agracey** [reply]
We could also sign the images with cosign so we know that they haven't been tampered with since we uploaded.

**Reply by agracey** [reply]
https://github.com/deislabs/bindle might be interesting to target once it's more ready. From my understanding, this would give the best of all worlds. But it's likely too early to adopt, so just noting it for awareness.

**Comment by agracey** [comment]
As a note for later: If we move to OCI, we will need to document any disaster recovery process.

**Comment by agracey** [comment]
For posterity: in a call yesterday, we decided to make this change in the not-to-distant future. 🎉 

It simplifies our deployment further which eases the maintenance burden for both us and the people administering our platform. It's also one less component to require knowledge for which improves the quality of our support.

**Comment by jimmykarily** [comment]
nice explanation of OCI artifacts: https://dlorenc.medium.com/oci-artifacts-explained-8f4a77945c13
